### DISCUSSION TITLE
Updating nested Helm values files where registry and repository information is separate from tag

### DISCUSSION THREAD

**Discussion by Cjewett** [renovatebot/renovate#24319] [Request Help]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

Testing locally using VSCode devcontainer flow.

### If you're self-hosting Renovate, select which platform you are using.

None

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

Nested values.yaml files specified in Helmfile release specification has the version updated.

### What you tried so far.

I created a repository to test all of this out here: https://github.com/Cjewett/renovate-helmfile-gotmpl-testing/tree/main/helmfile-testing 

I implemented some local changes:
helm-values index.ts altered to pick up values.yaml.gotmpl files.
helmfile index.ts altered to pick up helmfile.yaml.gotmpl files. 

A helmfile can specify multiple values.yaml files, so in some cases a 'parent' values file might contain the registry and repository information. https://github.com/Cjewett/renovate-helmfile-gotmpl-testing/blob/main/helmfile-testing/values_not_ok/values.yaml 

The underlying environment specific values.yaml files would specify the tag used in that environment. https://github.com/Cjewett/renovate-helmfile-gotmpl-testing/blob/main/helmfile-testing/values_not_ok/development/values.yaml 

The helmfile referencing these values.yaml is here: https://github.com/Cjewett/renovate-helmfile-gotmpl-testing/blob/main/helmfile-testing/helmfile.yaml.gotmpl

I tested this out quickly and it didn't bump the version for the values.yaml files which was expected. I imagine these files need at minimum the repository and tag specified. I was curious if the Renovate team has run into this nested use case and have any guidance. Is there functionality today that could get this to work or does it seem like more trouble than its worth?

### Relevant debug logs

<details><summary>Logs</summary>

```
Copy/paste the relevant log(s) here, between the starting and ending backticks
```

</details>

**Comment by rarkins** [comment]
Could you add one or more example, manual PRs demonstrating the type of updates which are today missing? You don't need to do the full Renovate PR body - we just care about the files.

**Reply by Cjewett** [reply]
Created an example of what we would update manually here: https://github.com/Cjewett/renovate-helmfile-gotmpl-testing/pull/12 

Edit: Not sure if this is exactly what you were asking for. If not could you clarify?

**Reply by rarkins** [reply]
Sorry for any silly questions as I'm not a regular Helm users. Is this problem about nested values, templates, or both?

When I see a file like https://github.com/Cjewett/renovate-helmfile-gotmpl-testing/blob/main/helmfile-testing/values_not_ok/development/values.yaml, my thought is "how is Renovate meant to know that this file is tied to something/somewhere else?" e.g is there a standard convention for such logic which is bulletproof?

**Reply by Cjewett** [reply]
@rarkins No problem and not silly at all!

In the case of Helmfiles which reference values.yaml you have to read the Helmfile, find each release, then for each release find all of the values.yaml files, and merge them to understand the context (registry, repository, tag). Last config wins if duplicates exist. After that you could make decisions to update the files. 

Admittedly it comes with a lot of baggage. The standard exists here https://helmfile.readthedocs.io/en/latest/#configuration but it's even further complicated by the fact that you have to render it w/ go templating, environment variables, etc. 

So to answer your question... I'd say it's both nested values and templating that makes this problem tricky. It all makes me lean towards telling people "just keep the image information in the same files always" but figured I'd at least reach out to start a discussion.

**Reply by rarkins** [reply]
Is the directory naming completely free choice? e.g. not restricted to fixed names like "development", etc? And is it always `values.yaml`?

**Reply by Cjewett** [reply]
Yeah directory naming is free choice. In our case we use go templating even with values.yaml, so the files would be values.yaml.gotmpl, but for most it would be values.yaml.
